std/experimental/checkedint.d

```diff
@@ -0,0 +1,2429 @@
+            if (lhs != rhs) return false;
+            // R(lhs) and R(rhs) have the same bit pattern, yet may be
+            // different due to signedness change.
+            static if (!isUnsigned!R)
+            {
+                if (rhs >= 0)
+                    return true;
+            }
+            else
+            {
+                if (lhs >= 0)
+                    return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+    Hook for `<`, `<=`, `>`, and `>=` that ensures comparison against integral
+    values has the behavior expected by the usual arithmetic rules. The built-in
+    semantics yield surprising behavior when comparing signed values against
+    unsigned values, for example $(D 0u < -1). The call $(D hookOpCmp(x, y))
+    returns `-1` if and only if `x` is smaller than `y` in abstract arithmetic
+    sense.
+
+    If one of the numbers is an integral and the other is a floating-point
+    number, $(D hookOpEquals(x, y)) returns a floating-point number that is `-1`
+    if `x < y`, `0` if `x == y`, `1` if `x > y`, and `NaN` if the floating-point
+    number is `NaN`.
+    */
+    static auto hookOpCmp(L, R)(L lhs, R rhs)
+    {
+        alias C = typeof(lhs + rhs);
+        static if (isFloatingPoint!C)
+        {
+            return lhs < rhs
+                ? C(-1)
+                : lhs > rhs ? C(1) : lhs == rhs ? C(0) : C.init;
+        }
+        else
+        {
+            static if (!valueConvertible!(L, C) || !valueConvertible!(R, C))
+            {
+                static assert(isUnsigned!C);
+                static assert(isUnsigned!L != isUnsigned!R);
+                if (!isUnsigned!L && lhs < 0)
+                    return -1;
+                if (!isUnsigned!R && rhs < 0)
+                    return 1;
+            }
+            return lhs < rhs ? -1 : lhs > rhs;
+        }
+    }
+}
+
+///
+unittest
+{
+    alias opEqualsProper = ProperCompare.hookOpEquals;
+    assert(opEqualsProper(42, 42));
+    assert(opEqualsProper(42u, 42));
+    assert(opEqualsProper(42, 42u));
+    assert(-1 == 4294967295u);
+    assert(!opEqualsProper(-1, 4294967295u));
+    assert(!opEqualsProper(uint(-1), -1));
+    assert(!opEqualsProper(uint(-1), -1.0));
+    assert(3_000_000_000U == -1_294_967_296);
+    assert(!opEqualsProper(3_000_000_000U, -1_294_967_296));
+}
+
+unittest
+{
+    alias opCmpProper = ProperCompare.hookOpCmp;
+    assert(opCmpProper(42, 42) == 0);
+    assert(opCmpProper(42u, 42) == 0);
+    assert(opCmpProper(42, 42u) == 0);
+    assert(opCmpProper(-1, uint(-1)) < 0);
+    assert(opCmpProper(uint(-1), -1) > 0);
+    assert(opCmpProper(-1.0, -1) == 0);
+}
+
+unittest
+{
+    auto x1 = Checked!(uint, ProperCompare)(42u);
+    assert(x1.get < -1);
+    assert(x1 > -1);
+}
+
+// WithNaN
+/**
+
+Hook that reserves a special value as a "Not a Number" representative. For
+signed integrals, the reserved value is `T.min`. For signed integrals, the
+reserved value is `T.max`.
```

Comment:
Typo: "For signed integrals, ... For signed integrals, ..." should be "For signed integrals, ... For unsigned integrals, ..."